test/test_fori_loop_with_while_loop_simple_add_dispatch_in_torch.py

```diff
@@ -19,88 +29,311 @@ def _fake_while_loop(cond_fn, body_fn, operands):
+    expected = _fake_while_loop_second(cond_fn, body_fn, (iteri, init_val, out_val))
+    self.assertEqual(res, expected)
 
   def test_while_loop_tpu_subtraction(self):
-
+    xm.mark_step()
     device = xm.xla_device()
 
-    def cond_fn(init, limit_value):
-      return limit_value[0] <= init[0]
+    def cond_fn(iteri, x):
+      return iteri > 0
+
+    def body_fn(iteri, x):
+      return iteri - 1, torch.sub(x, 1)
+
+    init_val = torch.tensor(10, dtype=torch.int32, device=device)
+    iteri = torch.tensor(3, device=device)
+    _, res = while_loop(cond_fn, body_fn, (iteri, init_val))
+    _, expected = _fake_while_loop_second(cond_fn, body_fn, (iteri, init_val))
+    self.assertTrue(torch.all(torch.eq(res, expected)))
+
+  def test_while_loop_tpu_addition_pure_torch(self):
+    def cond_fn(iteri, x, y):
+      return iteri > 0
 
-    def body_fn(init, limit_value):
-      one_value = torch.ones(1, dtype=torch.int32, device=device)
-      two_value = limit_value.clone()
-      return (torch.sub(init, one_value), two_value)
+    def body_fn(iteri, x, y):
+      return iteri - 1, x, torch.add(x, 1)
 
-    init = torch.tensor([10], dtype=torch.int32, device=device)
-    limit_value = torch.tensor([0], dtype=torch.int32, device=device)
-    res = while_loop(cond_fn, body_fn, (init, limit_value))
-    expected = _fake_while_loop(cond_fn, body_fn, (init, limit_value))
-    self.assertEqual(expected, res)
+    init_val = torch.tensor(3)
+    out_val = torch.tensor(15)
+    iteri = torch.tensor(10)
+    res =  while_loop(cond_fn, body_fn, (iteri, init_val, out_val))
+    expected = _fake_while_loop_second(cond_fn, body_fn, (iteri, init_val, out_val))
+    self.assertEqual(res, expected)
 
   def test_while_loop_tpu_addition(self):
+    xm.mark_step()
+    device = xm.xla_device()
+
+    def cond_fn(iteri, x):
+      return iteri > 0
+
+    def body_fn(iteri, x):
+      return iteri - 1, torch.add(x, 1)
+
+    init_val = torch.tensor(3, dtype=torch.int32, device=device)
+    iteri = torch.tensor(10, device=device)
+    _, res =  while_loop(cond_fn, body_fn, (iteri, init_val))
+    print("res: ", res)
+    _, expected = _fake_while_loop_second(cond_fn, body_fn, (iteri, init_val))
+    print("expected: ", expected)
+
+  def test_while_loop_tpu_addition_nested_pure_torch(self):
+
+    def cond_fn(iteri, x, y):
+      return iteri > 0
+
+    def body_fn(iteri, x, y):
+      return iteri - 1, x, torch.add(torch.add(x, 1), 1)
 
+    init_val = torch.tensor(0)
+    out_val = torch.tensor(0)
+    iteri = torch.tensor(10)
+    res =  while_loop(cond_fn, body_fn, (iteri, init_val, out_val))
+    expected = _fake_while_loop_second(cond_fn, body_fn, (iteri, init_val, out_val))
+    self.assertEqual(res, expected)
+
+  def test_while_loop_tpu_addition_nested(self):
+    xm.mark_step()
+    device = xm.xla_device()
+
+    def cond_fn(iteri, x):
+      return iteri > 0
+
+    def body_fn(iteri, x):
+      return iteri - 1, torch.add(torch.add(x, 1), 1)
```

Comment:
we don't need separate test for add and subtract, you can just keep one.

Reply:
thanks, updated